File: lib/src/main/java/com/dhy/retrofitrxutil/DisposableHandler.java
package com.dhy.retrofitrxutil;

import android.app.Application;
import android.content.Context;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.disposables.Disposable;

public final class DisposableHandler implements IDisposableHandler {
    private final Map<Context, List<Disposable>> map = new HashMap<>();

    @Override
    public void registerDisposable(@NonNull Context context, @NonNull Disposable disposable) {
        if (!(context instanceof Application)) {
            List<Disposable> list = map.get(context);
            if (list == null) {
                list = new ArrayList<>();
                map.put(context, list);
            }
            list.add(disposable);
        }
    }

    @Override
    public void onComplete(Context context, @NonNull Disposable disposable) {
        if (!(context instanceof Application)) {
            List<Disposable> list = map.get(context);
            if (list != null) list.remove(disposable);
        }
    }

    public void onDestroy(Context context) {
        List<Disposable> list = map.remove(context);
        if (list != null) for (Disposable disposable : list) {
            disposable.dispose();
        }
    }
}

File: lib/src/main/java/com/dhy/retrofitrxutil/DisposableHandler.kt
package com.dhy.retrofitrxutil

import android.app.Application
import android.content.Context

import io.reactivex.disposables.Disposable
import java.util.*

object DisposableHandler : IDisposableHandler {
    private val requests: WeakHashMap<Context, MutableList<Disposable>> = WeakHashMap()

    override fun registerDisposable(context: Context, disposable: Disposable) {
        if (context !is Application) {
            var list = requests[context]
            if (list == null) {
                list = ArrayList()
                requests[context] = list
            }
            list.add(disposable)
        }
    }

    override fun onComplete(context: Context, disposable: Disposable) {
        if (context !is Application) {
            val list = requests[context]
            list?.remove(disposable)
        }
    }

    override fun onDestroy(context: Context) {
        val dialog = DelayDialogProgress.dialogs[context]
        if (dialog?.isShowing == true) dialog.dismiss()

        val list = requests.remove(context)
        if (list != null) {
            val iterator = list.iterator()
            while (iterator.hasNext()) {
                iterator.next().dispose()
            }
        }
    }
}
